efficientdet/third_party/grad_checkpoint/memory_saving_gradients.py

```diff
@@ -1,21 +1,43 @@
+# Copyright 2020 Google Research. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# This file is a re-distribution based on the copyright in LICENSE.txt
+# ==============================================================================
 """Recompute tensorflow graph reducing GPU memory usage."""
-import logging
+# pylint: disable=g-direct-tensorflow-import,g-doc-args,g-doc-return-or-yield
+# pylint: disable=g-complex-comprehension,g-doc-exception
+# pylint: disable=protected-access
+
 import contextlib
-import time
+import logging
 import sys
+import time
 import numpy as np
 import tensorflow.compat.v1 as tf
 
-# save original gradients since tf.gradient could be monkey-patched to point
-# to our version
-from tensorflow.python.ops import gradients as tf_gradients_lib
-
-from third_party.toposort import toposort
+# save original gradients since tf.gradient could be monkey-patched.
 import third_party.graph_edit as ge
-
+from tensorflow.python.ops import gradients as tf_gradients_lib
 tf_gradient_function = tf_gradients_lib.gradients
 # logging.getLogger().setLevel(logging.DEBUG)
 
+
+def toposort(x):
+  # TODO(someone): should be the same as "from toposort import toposort".
```

Comment:
This is empty so far.  We should implement this function based on the latest commits in #716

Reply:
There is a toposort.py file in third_party. Do you want to merge it into memory_saving_gradients.py?

Reply:
Use a matrix table may accelerate toposort. 
Current implementation is based on adjacent table.